src/profile-logic/sanitize.js

```diff
@@ -496,41 +550,82 @@ function sanitizeThreadPII(
           }
         }
       }
+    }
 
-      // Now we'll remove samples related to the frames
-      // First we'll loop the stack table and populate a typed array with a flag
-      // indicating "this stack index has an ancestor with the private browsing
-      // flag". This means children inherits the private browsing flag. This is
-      // possible because when iterating we visit parents before their children.
-      const isStackPrivateBrowsing = new Uint8Array(stackTable.length);
-      for (let stackIndex = 0; stackIndex < stackTable.length; stackIndex++) {
-        const prefix = stackTable.prefix[stackIndex];
-        if (prefix !== null && isStackPrivateBrowsing[prefix]) {
-          isStackPrivateBrowsing[stackIndex] = 1;
+    // Now we'll remove samples related to the frames
+    const newSamples = (newThread.samples = {
+      ...samples,
+      stack: samples.stack.slice(),
+    });
+
+    // First we'll loop the stack table and populate a typed array with a value
+    // that is a flag that's inherited by children. This is possible because
+    // when iterating we visit parents before their children.
+    // There can be 3 values:
+    // - 0 is neutral, this means this stack isn't private browsing and isn't
+    //   the tab id we want to keep.
+    // - 1 means that this stack has at least one frame that's part of the tab
+    //   we want to keep (if any).
+    // - 2 means that this stack comes from private browsing. It always has
+    //   precedence on the active tab. This means that if the active tab comes
+    //   from a private browsing session and the user wants to sanitize it, it
+    //   will _still_ be removed.
+    // They won't be set if the related sanitization isn't set in PIIToBeRemoved.
+    // Also remember that one id can't be both in windowIdFromPrivateBrowsing
+    // and windowIdFromOtherTabs (windowIdFromPrivateBrowsing has precedence).
+    const stackFlags = new Uint8Array(stackTable.length);
```

Comment:
This array implements the "inheritance" behavior. As a side note, it could be easier to have `innerWindowID` inherited at processing time, just like we do for categories.
This would probably make `filterThreadByTab` a lot simpler too!